common/src/linux/xsave_gnu.S

```diff
@@ -89,7 +89,7 @@ DECLARE_LOCAL_FUNC restore_xregs
     lea     g_xsave_enabled, %eax
 #else
     mov     %rdi, %rcx
-    lea_pic g_xsave_enabled, %rax
```

Comment:
The code in line 72/73 is to load the address, then the code in line 75/76 is to load the actual value.

Reply:
After poking more at asm guides (I am _far_ from x86 asm proficient) it appears you are correct.

However, what we observed is that the g_xsave_enabled flag is always evaluated to "true" which is breaking us on older platforms. If we run sim on Haswell, the sim will crash every time when it runs DO_XRSTOR. We do _not_ see this on Skylake.

yuyuany@, in a separate message (I cannot seem to find it in github's UI) seemed to indicate the real problem is symbol confusion?
yuyuany@ said:
> Yes, that's a bug. g_xsave_enabled, g_xsave_mask_low, and g_xsave_mask_high are
> defined in both urts and trts. In HW mode, urts is a shared library, and there's no problem.
> But in simulation mode, urts is a static library as trts. lea_pic loads the symbol address
> from urts, rather than trts. I'll submit another PR to fix it.

If my proposed fix is wrong, I'm happy to close out this PR and await a real fix. Any ideas when it might be available?

Reply:
I checked issue again. There's no problem in both HW and simulation mode. The lea instruction can get the correct symbol address from enclave. 
Could you provide me the OS and GCC version you are using? Maybe it fails on some specific GCC/distro version.

Reply:
I've been digging into this more on our end, and I think we have some issues with our toolchain. We have our own build scripts, effectively (based on bazel, not make). I think that's the heart of the problem.

Most likely, I'll revert this PR, but I am going to hold onto it until we get to the bottom of the toolchain issues on our end.

Thanks for the assistance!

Reply:
I've done a lot of digging and manual reading :)

It turns out two things are going on here:
1. We do have a toolchain difference in asylo vs linux-sgx open source. What's significant is that we sometimes link xsave_gnu.o into a shared object. This does not work with `lea g_xsave_enabled(%rip), reg`, because `g_xsave_enabled` might require relocation.
2. Somebody attempted to "fix" this on our end by adding `@GOTPCREL` to the `lea_pic` macro, which broke everything. I was not aware of this previously.

Thus, I think this patch still makes a lot of sense to keep. Consider the case of some consumer of this code (such as Asylo) that links xsave_gnu.o into a shared object. When that happens, symbols like g_xsave_enabled need to be looked up in the global offset table. Thus, `mov g_xsave_enabled@GOTPCREL(%rip), reg` is more correct for that relocatable symbol. If we try to link the LEA version of the code, the result is an error: `xsave_gnu.pic.o: requires dynamic R_X86_64_PC32 reloc against 'g_xsave_enabled' which may overflow at runtime`

Lastly, if static linking is used, the linker is smart enough to replace the MOV with LEA. The resulting bytecode with a static link is `lea offset(%rip), reg`. However, linking a shared object results in `mov offset(%rip), reg`.

In light of that, would you please reconsider accepting this patch?

Reply:
We did see some issue with "mov" on some OS/toolchain, so that's why it was changed to "lea".  Accepting this patch means it will back to the previous situation.  

BTW, the code is used in shared library (both Enclave and ligsgx_urts_sim.so are shared libraries, not static libraries), so it is strange to fail in your shared objects.

Reply:
Also this code should support locating the address before and after relocation, since we have some other places using the same "lea" approach to locate the variable even before relocation.

Reply:
What I saw with our toolchain is that, when the variable can be located local to the library, the linker emits `lea offset($rip), %rax`. In fact, it doesn't matter if the macro assembly code has lea_pic or the old move--the output (according to objdump) is identical. I found this pretty interesting, and it seems like common behavior.

However, when we build a shared library, the linker emits a MOV which references an entry in the global offset table. Our toolchain refuses to build shared libraries if the input macro assembly has lea_pic in it.

Asylo builds static libraries for almost everything, including urts_sim. We only have a few .sos, and it's those that have problems when linking urts_sim. Asylo already maintains a patch for compatibility with our toolchain, so if the LEA changes were made to fix other toolchains, I think we can continue to maintain our patch. We were just hoping to have one less diff between our code and the mainline.

tl;dr--we can keep a fork in Asylo if you believe this patch is not appropriate for upstreaming.

Reply:
If urts_sim is a static library, it might have issue that enclave is referencing the g_xsave_enabled in urts_sim (since they are linked together?), which is not appropriate because it is supposed to access the same named variable in trts_sim.   Or because of this implementation, there is only one g_xave_enabled variable?

Reply:
The enclave itself shouldn't be referencing urts_sim, as far as I'm aware. The enclave will link trts_sim. Thus, I think we're fine because there is only one g_xsave_enabled variable.

I'd have to double check the Asylo build rules, but I believe we're careful to swap between urts/trts depending on whether code is outside/inside an enclave.